### PR TITLE
refactor(lsp): always crop matching text from end of incremental sync range

### LINKED ISSUE #16624: TexLab error when inserting text before multibyte characters
### Neovim version (nvim -v)

v0.7.0-dev+714-g3aff3d634, also exists in v0.6.0

### Language server name/version

texlab 3.3.1

### Operating system/version

Arch Linux

### :checkhealth

```markdown
lspconfig: require("lspconfig.health").check()
========================================================================
  - INFO: texlab: configuration checked.

nvim: health#nvim#check
========================================================================
## Configuration
  - WARNING: Missing user config file: /home/zeertzjq/.config/nvim/init.vim
    - ADVICE:
      - :help |init.vim|

## Performance
  - INFO: Build type: Debug
  - WARNING: Non-optimized (DEBUG) build. Nvim will be slower.
    - ADVICE:
      - Install a different Nvim package, or rebuild with `CMAKE_BUILD_TYPE=RelWithDebInfo`.
      - https://github.com/neovim/neovim/wiki/FAQ

## Remote Plugins
  - OK: Up to date

## terminal
  - INFO: key_backspace (kbs) terminfo entry: key_backspace=\177
  - INFO: key_dc (kdch1) terminfo entry: key_dc=\E[3~
  - INFO: $COLORTERM='truecolor'
```

### Steps to reproduce using "nvim -u minimal_init.lua"

This is `/tmp/minimal_init.lua`:
```lua
vim.lsp.set_log_level('trace')
require('vim.lsp.log').set_format_func(vim.inspect)
vim.cmd([[set rtp+=~/.local/share/nvim/site/pack/packer/start/nvim-lspconfig]])
require('lspconfig').texlab.setup({})
vim.o.showmode = false
```

This is `/tmp/tmp.tex`:
```tex
\begin{document}
→
\end{document}
```

1. Run `VIMRUNTIME=./runtime ./build/bin/nvim --clean -u /tmp/minimal_init.lua /tmp/tmp.tex`
2. Press `2ggII`

### Expected behavior

No errors. LSP client doesn't quit.

### Actual behavior

`Client 1 quit with exit code 1 and signal 0`

### Log file

https://gist.github.com/zeertzjq/5afb01d99afd99af9537e12669b05905

### PR DESCRIPTION
**Pull request by rish987**
This seems like a much more accurate way of doing things here, but let me know if I missed anything. Added a number of tests, but I haven't been doing a lot of Lean 4 recently so @gebner some beta-testing would be appreciated.

### CODE REVIEW

**Comment by mjlbach** [comment]
The tests are fine, I'd rather avoid unnecessary changes. This is more similar to more original implementation in https://github.com/neovim/neovim/commit/2ecf0a4c6183bba1c65d440711038f040d355fef#diff-e0bb627783152fce1040c4ecdbd81d6c178095a1f4ceeca33007b86249437434R87-R118 which used a unified align function which I removed to simplify the code.

File: runtime/lua/vim/lsp/sync.lua [hunk 1]
```diff
@@ -93,30 +93,35 @@ end
...
+        byte = byte + offset
+      else
+        byte = byte + str_utf_end(line, byte) + 1
+      end
+    end
+  end
+
   local char
-  -- If on the first byte, or an empty string: the trivial case
-  if byte == 1 or #line == 0 then
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/sync.lua:101]: Why did you remove this optimization?

File: test/functional/plugin/lsp/incremental_sync_spec.lua [hunk 2]
```diff
@@ -265,12 +460,12 @@ describe('incremental synchronization', function()
               line = 0
             },
             ['end'] = {
-              character = 17,
+              character = 12,
               line = 0
             }
           },
-          rangeLength = 6,
-          text = '\ntest3'
```
Comment (mjlbach) [inline test/functional/plugin/lsp/incremental_sync_spec.lua:273]: This test change looks problematic

File: runtime/lua/vim/lsp/sync.lua [hunk 3]
```diff
@@ -209,51 +203,48 @@ local function compute_end_range(prev_lines, curr_lines, start_range, firstline,
...
+  -- Maximum number of bytes to search backwards for mismatch
+  local max_length = min(prev_line_range, curr_line_range)
 
-  -- Editing the same line
-  -- If the byte offset is zero, that means there is a difference on the last byte (not newline)
-  if prev_line_idx == curr_line_idx then
-    local max_length
-    if start_line_idx == prev_line_idx then
-      -- Search until beginning of difference
-      max_length = min(prev_line_length - start_range.byte_idx, curr_line_length - start_range.byte_idx) + 1
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/sync.lua:220]: Why are you removing this? We should not search if the lines are the same beyond the first change of the line.

Comment (rish987) [reply runtime/lua/vim/lsp/sync.lua:220]: See above, I've set these individually for `prev_line` and `curr_line`.

Comment (rish987) [reply runtime/lua/vim/lsp/sync.lua:220]: Lines 208, 214.

File: runtime/lua/vim/lsp/sync.lua (hunk 1 above)
Comment (rish987) [reply runtime/lua/vim/lsp/sync.lua:101]: My bad, just added an equivalent optimization for this case?

File: test/functional/plugin/lsp/incremental_sync_spec.lua (hunk 2 above)
Comment (rish987) [reply test/functional/plugin/lsp/incremental_sync_spec.lua:273]: How so? It seems we're just replacing a space with a newline here.

**Comment by mjlbach** [comment]
To put this more strongly. I'm not merging this as-is. Fix the issue without stylistically refactoring sync.lua, and in general please do not introduce refactors and fixes in the same PR. It makes it extremely difficult to review, and the fix is not documented anywhere.

File: runtime/lua/vim/lsp/sync.lua (hunk 1 above)
Comment (mjlbach) [reply runtime/lua/vim/lsp/sync.lua:101]: This isn't an equivalent optimization. It doesn't short-ircuit the valuation as it did previously. You're still checking the byte offset.

Comment (rish987) [reply runtime/lua/vim/lsp/sync.lua:101]: Good catch, just fixed.

Comment (mjlbach) [reply runtime/lua/vim/lsp/sync.lua:101]: Hmm is there not a scenario in which byte > #line and #line == 0 that this is not catching?

**Comment by rish987** [comment]
> in general please do not introduce refactors and fixes in the same PR.

I'll see if I can get something minimal together for 0.6.1. Really my intention was for this to be for 0.7.0 after a good amount of review and beta testing.

File: runtime/lua/vim/lsp/sync.lua (hunk 1 above)
Comment (rish987) [reply runtime/lua/vim/lsp/sync.lua:101]: In this case `char` should be 1 right?

Comment (mjlbach) [reply runtime/lua/vim/lsp/sync.lua:101]: In the current implementation, that is the case

Comment (rish987) [reply runtime/lua/vim/lsp/sync.lua:101]: It should be the case here as well. In `compute_end_range()`, if one of the lines has length 0 the `byte_offset` stays at -1. (`compute_start_range()` wasn't really changed)

**Comment by rish987** [comment]
See #16670

File: runtime/lua/vim/lsp/sync.lua [hunk 4]
```diff
@@ -210,51 +206,48 @@ local function compute_end_range(prev_lines, curr_lines, start_range, firstline,
...
-  if prev_end_byte_idx == 0 then
-    prev_end_byte_idx = 1
-  end
-  local prev_byte_idx, prev_char_idx = align_end_position(prev_line, prev_end_byte_idx, offset_encoding)
+  local prev_byte_idx, prev_char_idx = align_position(prev_line, prev_end_byte_idx, false, offset_encoding)
   local prev_end_range = { line_idx = prev_line_idx, byte_idx = prev_byte_idx, char_idx = prev_char_idx }
 
-  local curr_end_range
-  -- Deletion event, new_range cannot be before start
-  if curr_line_idx < start_line_idx then
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/sync.lua:247]: Did you remove the explicit handling of this?

**Comment by mjlbach** [comment]
perf -> refactor (unless you have benchmarks :))

File: runtime/lua/vim/lsp/sync.lua [hunk 5]
```diff
@@ -93,31 +93,38 @@ end
...
   end
+
+  local char
+
+  -- optimize for first byte case
+  if byte == 1 then
+    char = 1
+  -- Called in the case of extending an empty line "" -> "a"
+  elseif byte == #line + 1 then
+    char = compute_line_length(line, offset_encoding) + 1
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/sync.lua:123]: These functions can be easily unified since I allowed byteindex to take explicit nuls

File: runtime/lua/vim/lsp/sync.lua (hunk 4 above)
Comment (rish987) [reply runtime/lua/vim/lsp/sync.lua:247]: This check didn't make sense to me. AFAICT the only way this can arise is if `firstline == new_lastline`, which we checked at the very beginning of this function.

File: runtime/lua/vim/lsp/sync.lua (hunk 5 above)
Comment (rish987) [reply runtime/lua/vim/lsp/sync.lua:123]: See #16731, we should be able to just use those helpers directly.

Comment (rish987) [reply runtime/lua/vim/lsp/sync.lua]: ^ done, let me know what you think

**Comment by mjlbach** [comment]
I am now fine with this, I suggest daily driving this for several weeks and i will merge between 0.6.1 and 0.6.2.

**Comment by rish987** [comment]
Will do, though @gebner is probably a better daily driver than me for this (if you don't mind since I'm not doing any Lean 4 yet).

**Comment by gebner** [comment]
FWIW, I've been running this patch for a while now and I haven't run into any synchronization-related errors.

**Comment by mjlbach** [comment]
I'm going to wait on this one until I've fixed https://github.com/neovim/neovim/issues/17085

**Comment by mfussenegger** [comment]
@rish987 is this PR still relevant, and if so can you outline a concrete issue that it would fix?

**Comment by rish987** [comment]
Hey, it's been quite a while hahaha. IIRC this was mainly just a refactor of the way we send document change updates to the server. It's perhaps worth mentioning to a maintainer if they ever want to overhaul this function, but for now it's probably not worth our attention, so I'll close it off.
